File: tests/scc_yosupo.test.py
# verify-helper: PROBLEM https://judge.yosupo.jp/problem/scc
# @import python_library/graph/scc.py
import sys

sys.path.insert(0, ".")
input = sys.stdin.buffer.readline
sys.setrecursionlimit(6 * 10 ** 5)

from python_library.graph.scc import StronglyConnectedComponets


def main() -> None:
    N, M = map(int, input().split())
    scc = StronglyConnectedComponets(N)
    for _ in range(M):
        u, v = map(int, input().split())
        scc.add_edge(u, v)
    scc.run()
    ans_size = max(scc.order) + 1
    ans = [[] for _ in range(ans_size)]
    for i, c in enumerate(scc.order):
        ans[c].append(i)
    print(ans_size)
    for row in ans:
        print(len(row), *row)


if __name__ == "__main__":
    main()

File: tests/cpp/scc_yosupo.test.cpp
#define PROBLEM "https://judge.yosupo.jp/problem/scc"
#include <iostream>
#include <map>

#include "c++-library/scc.cpp"

void yosupo() {
  int N, M;
  std::cin >> N >> M;
  StronglyConnectedComponents scc(N);
  for (int i = 0; i < M; i++) {
    int a, b;
    std::cin >> a >> b;
    scc.add_edge(a, b);
  }
  scc.run();
  std::map<int, std::vector<int>> ans;
  for (int i = N - 1; i >= 0; i--) {
    ans[scc.cmp[i]].emplace_back(i);
  }
  std::cout << ans.size() << std::endl;
  for (auto&& p : ans) {
    std::cout << p.second.size();
    for (int& i : p.second) std::cout << ' ' << i;
    std::cout << std::endl;
  }
}

int main() {
  std::cin.tie(0);
  std::ios_base::sync_with_stdio(false);

  yosupo();
  return 0;
}